README.md
# cookiecutter-git

[![Tagged Release](https://img.shields.io/badge/release-v0.6.1-blue.svg?longCache=true)](https://github.com/NathanUrwin/cookiecutter-git/releases/tag/v0.6.1)
[![Development Status](https://img.shields.io/badge/status-beta-brightgreen.svg?longCache=true)](ROADMAP.md)
[![Build Status](https://travis-ci.org/NathanUrwin/cookiecutter-git.svg?branch=master)](https://travis-ci.org/NathanUrwin/cookiecutter-git)
[![Build Status](https://ci.appveyor.com/api/projects/status/3q0aik8sgmndwibp/branch/master?svg=true)](https://ci.appveyor.com/project/NathanUrwin/cookiecutter-git/branch/master)
[![codecov](https://codecov.io/gh/NathanUrwin/cookiecutter-git/branch/master/graph/badge.svg)](https://codecov.io/gh/NathanUrwin/cookiecutter-git)
[![python](https://img.shields.io/badge/python-2.7%2C%203.4%2C%203.5%2C%203.6%2C%203.7-blue.svg?longCache=true)](https://www.python.org/downloads/)

> Git repo project template using Cookiecutter :cookie:

[![Cookiecutter-Git Logo](images/logo-256.png)](https://dylantyates.com/graphics)

This project is inspired by [cookiecutter-template by eviweb](https://github.com/eviweb/cookiecutter-template), and consists of a [cookiecutter](https://github.com/audreyr/cookiecutter#cookiecutter) (project template) that provides the necessary [markdown](https://guides.github.com/features/mastering-markdown/) docs and other files to pass [GitHub's open source guidelines](https://opensource.guide/) with an added bonus: remote repos are created for you [*automagically*](https://youtu.be/Z3qK8gT5LLg?t=24s)! :crystal_ball::zap::boom:

_**Note:** The [Beta release](https://github.com/NathanUrwin/cookiecutter-git/releases/tag/v0.5.0) introduces [breaking changes](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.4.1...v0.5.0)! [Invoke](http://docs.pyinvoke.org/en/1.1/) and [Requests](http://docs.python-requests.org/en/master/) are **now required** and [the prompts](#documentation) have changed!_

## Table of Contents

- [Table of Contents](#table-of-contents)
- [Features](#features)
- [Requirements](#requirements)
  - [Recommended](#recommended)
- [Installation](#installation)
  - [Users](#users)
  - [Contributors](#contributors)
- [Usage](#usage)
  - [Example](#example)
- [Documentation](#documentation)
- [Resources](#resources)
- [Development](#development)
  - [Future](#future)
  - [History](#history)
  - [Community](#community)
- [Credits](#credits)
- [License](#license)

## Features

- Creates new project directory
- [Bare project structure](https://github.com/nathanurwin/cookiecutter-git-demo)
  - For any programming language or codebase
  - Useful but not overruling organization
- [License customization](https://choosealicense.com/)
- Code of Conduct customization
- [Git Ignore customization](https://www.gitignore.io/)
- Git repo initialization
- Remote repo creation
  - [Bitbucket.org](https://bitbucket.org/) using HTTP Basic auth (2FA disabled only)
  - [GitHub.com](https://github.com/) using HTTP Basic auth (2FA support *coming soon*)
  - [GitLab.com](https://gitlab.com/) using `git push` (HTTP Basic auth or SSH)
- Cross-platform support
  - [Windows](https://www.microsoft.com/en-us/windows)
  - [macOS](https://www.apple.com/macos/high-sierra/)
  - [Linux](https://www.linux.org/)

## Requirements

- [Cookiecutter](https://github.com/audreyr/cookiecutter)
- [Invoke](http://www.pyinvoke.org/)
- [Requests](http://docs.python-requests.org/en/master/)

_**Note:** Cookiecutter **should be** installed with [pip](https://pip.pypa.io/en/stable/installing/), or else `invoke` and `requests` may not be in `$PATH` and/or `$PYTHONPATH`!_

### Recommended

- [ghi](https://github.com/stephencelis/ghi)
- [github-changelog-generator](https://github.com/github-changelog-generator/github-changelog-generator)
- [pyenv](https://github.com/pyenv/pyenv) ([conda](https://github.com/conda/conda) on Windows)

## Installation

### Users

```bash
$ pip install --user cookiecutter invoke requests
Collecting cookiecutter
  Using cached https://files.pythonhosted.org/packages/16/99/1ca3a75978270288354f419e9166666801cf7e7d8df984de44a7d5d8b8d0/cookiecutter-1.6.0-py2.py3-none-any.whl
Collecting invoke
  Using cached https://files.pythonhosted.org/packages/6c/66/9e232c59e61f0a0b6552d68419a5c5a5dba368e105fdbfd2b6c74c402234/invoke-1.0.0-py3-none-any.whl
Collecting requests
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/65/47/7e02164a2a3db50ed6d8a6ab1d6d60b69c4c3fdf57a284257925dfc12bda/requests-2.19.1-py2.py3-none-any.whl
Collecting binaryornot>=0.2.0 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/24/7e/f7b6f453e6481d1e233540262ccbfcf89adcd43606f44a028d7f5fae5eb2/binaryornot-0.4.4-py2.py3-none-any.whl
Collecting poyo>=0.1.0 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/ea/6c/62c76c12015f6a1849446fb73da59be1229312c54d6d05068275e52bf29f/poyo-0.4.1-py2.py3-none-any.whl
Collecting future>=0.15.2 (from cookiecutter)
Collecting jinja2-time>=0.1.0 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/6a/a1/d44fa38306ffa34a7e1af09632b158e13ec89670ce491f8a15af3ebcb4e4/jinja2_time-0.2.0-py2.py3-none-any.whl
Collecting whichcraft>=0.4.0 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/60/8a/5c52e30e11672f7e3aa61f348ddae443d122bcd96bc8b785ac76dbae944b/whichcraft-0.4.1-py2.py3-none-any.whl
Collecting jinja2>=2.7 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/7f/ff/ae64bacdfc95f27a016a7bed8e8686763ba4d277a78ca76f32659220a731/Jinja2-2.10-py2.py3-none-any.whl
Collecting click>=5.0 (from cookiecutter)
  Using cached https://files.pythonhosted.org/packages/34/c1/8806f99713ddb993c5366c362b2f908f18269f8d792aff1abfd700775a77/click-6.7-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests)
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/7c/e6/92ad559b7192d846975fc916b65f667c7b8c3a32bea7372340bfe9a15fa5/certifi-2018.4.16-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests)
  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl
Collecting urllib3<1.24,>=1.21.1 (from requests)
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/bd/c9/6fdd990019071a4a32a5e7cb78a1d92c53851ef4f56f62a3486e6a7d8ffb/urllib3-1.23-py2.py3-none-any.whl
Collecting idna<2.8,>=2.5 (from requests)
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/4b/2a/0276479a4b3caeb8a8c1af2f8e4355746a97fab05a372e4a2c6a6b876165/idna-2.7-py2.py3-none-any.whl
Collecting arrow (from jinja2-time>=0.1.0->cookiecutter)
Collecting MarkupSafe>=0.23 (from jinja2>=2.7->cookiecutter)
Collecting python-dateutil (from arrow->jinja2-time>=0.1.0->cookiecutter)
  Cache entry deserialization failed, entry ignored
  Using cached https://files.pythonhosted.org/packages/cf/f5/af2b09c957ace60dcfac112b669c45c8c97e32f94aa8b56da4c6d1682825/python_dateutil-2.7.3-py2.py3-none-any.whl
Collecting six>=1.5 (from python-dateutil->arrow->jinja2-time>=0.1.0->cookiecutter)
  Using cached https://files.pythonhosted.org/packages/67/4b/141a581104b1f6397bfa78ac9d43d8ad29a7ca43ea90a2d863fe3056e86a/six-1.11.0-py2.py3-none-any.whl
Installing collected packages: chardet, binaryornot, poyo, future, six, python-dateutil, arrow, MarkupSafe, jinja2, jinja2-time, whichcraft, certifi, urllib3, idna, requests, click, cookiecutter, invoke
Successfully installed MarkupSafe-1.0 arrow-0.12.1 binaryornot-0.4.4 certifi-2018.4.16 chardet-3.0.4 click-6.7 cookiecutter-1.6.0 future-0.16.0 idna-2.7 invoke-1.0.0 jinja2-2.10 jinja2-time-0.2.0 poyo-0.4.1 python-dateutil-2.7.3 requests-2.19.1 six-1.11.0 urllib3-1.23 whichcraft-0.4.1
```

See [Usage](#usage)

### Contributors

```bash
$ git clone https://github.com/NathanUrwin/cookiecutter-git
Cloning into 'cookiecutter-git'...
remote: Counting objects: 757, done.
remote: Compressing objects: 100% (80/80), done.
remote: Total 757 (delta 63), reused 73 (delta 32), pack-reused 645
Receiving objects: 100% (757/757), 337.48 KiB | 0 bytes/s, done.
Resolving deltas: 100% (432/432), done.
$ cd cookiecutter-git
$ pipenv install --dev --pre
Creating a virtualenv for this project…
Using /home/user/.pyenv/versions/3.6.6/bin/python3.6m to create virtualenv…
⠋Running virtualenv with interpreter /home/user/.pyenv/versions/3.6.6/bin/python3.6m
Using base prefix '/home/user/.pyenv/versions/3.6.6'
New python executable in /home/user/.local/share/virtualenvs/cookiecutter-git-Jx1W2Sde/bin/python3.6m
Also creating executable in /home/user/.local/share/virtualenvs/cookiecutter-git-Jx1W2Sde/bin/python
Installing setuptools, pip, wheel...done.

Virtualenv location: /home/user/.local/share/virtualenvs/cookiecutter-git-Jx1W2Sde
Installing dependencies from Pipfile.lock (25c20f)…
  🐍   ▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉▉ 58/58 — 00:00:07
To activate this project's virtualenv, run the following:
 $ pipenv shell
$ pipenv run invoke tests
# before running tests you may have to run:
$ pipenv run pip install -r requirements.txt
```

See [CONTRIBUTING](#contributing)

## Usage

```bash
$ mkdir -p ~/Projects/NathanUrwin
$ cd ~/Projects/NathanUrwin
$ cookiecutter gh:NathanUrwin/cookiecutter-git
You've downloaded /home/user/.cookiecutters/cookiecutter-git before. Is it okay to delete and re-download it? [yes]:
git_name [Nathan Urwin]:
git_email [me@nathanurwin.com]:
git_ignore [windows,macos,linux,git]:
repo_slug [cookiecutter-git-demo]:
repo_tagline [A cookiecutter-git demonstration :tada:]:
repo_summary [This project ...]:
Select remote_provider:
1 - github.com
2 - gitlab.com
3 - bitbucket.org
4 - none
Choose from 1, 2, 3, 4 [1]:
remote_username [NathanUrwin]:
remote_namespace [NathanUrwin]:
Select remote_protocol:
1 - https
2 - ssh
Choose from 1, 2 [1]:
Select code_of_conduct:
1 - contributor_covenant
2 - citizen_code_of_conduct
Choose from 1, 2 [1]:
Select copyright_license:
1 - MIT
2 - Apache-2.0
3 - BSD-2-Clause
4 - BSD-3-Clause
5 - GPL-2.0
6 - GPL-3.0
7 - AGPL-3.0
8 - LGPL-2.1
9 - LGPL-3.0
10 - EPL-1.0
11 - MPL-2.0
12 - Unlicense
Choose from 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12 [1]:
copyright_holder [Nathan Urwin]:
make_dirs [docs,src,tests]:
Password for 'https://NathanUrwin@github.com':
Initialized empty Git repository in /home/user/Projects/NathanUrwin/cookiecutter-git-demo/.git/
[master (root-commit) 5bdaa5c] Initial commit
 15 files changed, 350 insertions(+)
 create mode 100644 .editorconfig
 create mode 100644 .github/ISSUE_TEMPLATE/bug_report.md
 create mode 100644 .github/ISSUE_TEMPLATE/feature_request.md
 create mode 100644 .github/PULL_REQUEST_TEMPLATE.md
 create mode 100644 .gitignore
 create mode 100644 AUTHORS.md
 create mode 100644 CHANGELOG.md
 create mode 100644 CODE_OF_CONDUCT.md
 create mode 100644 CONTRIBUTING.md
 create mode 100644 LICENSE
 create mode 100644 README.md
 create mode 100644 ROADMAP.md
 create mode 100644 docs/.gitkeep
 create mode 100644 src/.gitkeep
 create mode 100644 tests/.gitkeep
To https://github.com/NathanUrwin/cookiecutter-git-demo.git
 * [new branch]      master -> master
Branch master set up to track remote branch master from origin.


Success! Your project was created here:
/home/user/Projects/NathanUrwin/cookiecutter-git-demo
Also see: https://github.com/NathanUrwin/cookiecutter-git-demo
Thanks for using cookiecutter-git! :)


```

See generated [README.md](https://github.com/NathanUrwin/cookiecutter-git-demo/blob/master/README.md)

### Example

See [cookiecutter-git-demo](https://github.com/NathanUrwin/cookiecutter-git-demo)

```bash
$ tree -a -I .git cookiecutter-git-demo
cookiecutter-git-demo
├── AUTHORS.md
├── CHANGELOG.md
├── CODE_OF_CONDUCT.md
├── CONTRIBUTING.md
├── docs
│   └── .gitkeep
├── .editorconfig
├── .github
│   ├── ISSUE_TEMPLATE
│   │   ├── bug_report.md
│   │   └── feature_request.md
│   └── PULL_REQUEST_TEMPLATE.md
├── .gitignore
├── LICENSE
├── README.md
├── ROADMAP.md
├── src
│   └── .gitkeep
└── tests
    └── .gitkeep

5 directories, 15 files
```

## Documentation

Cookiecutter prompts explained in-depth. See [cookiecutter.json](cookiecutter.json) for default values.

Prompt | Explanation
--- | ---
`git_name` | Your full name, including first and last names, titles, and possibly even your middle name. This will go under *Core Contributor* in **AUTHORS.md**. See `git config --global user.name`
`git_email` | Your git user email address you want associated with the repository. This will go under *Core Contributor* in **AUTHORS.md**. See `git config --global user.email`
`git_ignore` | A comma-separated values (csv) list of preset templates of paths for git to ignore. See the [gitignore.io README](https://github.com/joeblau/gitignore.io#list) for available values. This will be used to generate the **.gitignore** file.
`repo_slug` | The repository name containing only alphanumeric characters and dashes. This will be the local, top-level directory name, the remote repo endpoint, and the *H1* in the **README.md**.
`repo_tagline` | A short description about the repository in *50 words or less*. This will be the remote description setting, and the content under the *H1* in the **README.md**.

CHANGELOG.md
# Change Log

## [v0.6.1](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.6.1) (2018-08-06)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.6.0...v0.6.1)

**Implemented enhancements:**

- Add note about $PATH [\#35](https://github.com/NathanUrwin/cookiecutter-git/issues/35)

**Fixed bugs:**

- Fix Windows support [\#42](https://github.com/NathanUrwin/cookiecutter-git/issues/42)

## [v0.6.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.6.0) (2018-08-01)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.5.1...v0.6.0)

**Implemented enhancements:**

- Add code\_of\_conduct prompt [\#38](https://github.com/NathanUrwin/cookiecutter-git/issues/38)
- Add AppVeyor support [\#33](https://github.com/NathanUrwin/cookiecutter-git/issues/33)

**Fixed bugs:**

- Fix repo\_slug dirs outdated content [\#39](https://github.com/NathanUrwin/cookiecutter-git/issues/39)
- Remove copy\_cookiecutter\_git prompt [\#37](https://github.com/NathanUrwin/cookiecutter-git/issues/37)
- Fix tests on Windows [\#26](https://github.com/NathanUrwin/cookiecutter-git/issues/26)

## [v0.5.1](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.5.1) (2018-07-30)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.5.0...v0.5.1)

**Fixed bugs:**

- Fix Invoke on Windows [\#34](https://github.com/NathanUrwin/cookiecutter-git/issues/34)

## [v0.5.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.5.0) (2018-07-26)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.4.1...v0.5.0)

**Implemented enhancements:**

- Add cookie-cookie feature [\#32](https://github.com/NathanUrwin/cookiecutter-git/issues/32)
- Add Codecov support [\#30](https://github.com/NathanUrwin/cookiecutter-git/issues/30)
- Add Travis Continuous Integration support [\#9](https://github.com/NathanUrwin/cookiecutter-git/issues/9)

**Fixed bugs:**

- Fix gitlab.com as remote\_provider [\#22](https://github.com/NathanUrwin/cookiecutter-git/issues/22)
- Fix urllib2 URLError urlopen error [\#21](https://github.com/NathanUrwin/cookiecutter-git/issues/21)
- Fix bash tests with pytest [\#8](https://github.com/NathanUrwin/cookiecutter-git/issues/8)

## [v0.4.1](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.4.1) (2018-03-16)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.4.0...v0.4.1)

**Fixed bugs:**

- Bitbucket remote repo failed [\#20](https://github.com/NathanUrwin/cookiecutter-git/issues/20)

## [v0.4.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.4.0) (2018-03-15)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.3.0...v0.4.0)

**Implemented enhancements:**

- Add python3 support [\#15](https://github.com/NathanUrwin/cookiecutter-git/issues/15)

**Fixed bugs:**

- Fix UnboundLocalError [\#17](https://github.com/NathanUrwin/cookiecutter-git/issues/17)
- Fix Windows TypeError [\#16](https://github.com/NathanUrwin/cookiecutter-git/issues/16)

## [v0.3.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.3.0) (2017-03-25)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.2.0...v0.3.0)

**Implemented enhancements:**

- Add cookiecutter.json documentation [\#14](https://github.com/NathanUrwin/cookiecutter-git/issues/14)
- Refactor cookiecutter.json keys [\#13](https://github.com/NathanUrwin/cookiecutter-git/issues/13)

## [v0.2.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.2.0) (2017-03-24)

[Full Changelog](https://github.com/NathanUrwin/cookiecutter-git/compare/v0.1.0...v0.2.0)

**Implemented enhancements:**

- Add bitbucket.org support [\#5](https://github.com/NathanUrwin/cookiecutter-git/issues/5)
- Add first tagged release [\#3](https://github.com/NathanUrwin/cookiecutter-git/issues/3)

## [v0.1.0](https://github.com/NathanUrwin/cookiecutter-git/tree/v0.1.0) (2017-03-24)

**Implemented enhancements:**

- Add more license choices [\#4](https://github.com/NathanUrwin/cookiecutter-git/issues/4)
- Use github-changelog-generator [\#2](https://github.com/NathanUrwin/cookiecutter-git/issues/2)
- Add gitlab.com support [\#1](https://github.com/NathanUrwin/cookiecutter-git/issues/1)

**Fixed bugs:**

- Fix insecure gitlab\_token retrieval [\#6](https://github.com/NathanUrwin/cookiecutter-git/issues/6)

\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


hooks/post_gen_project.py
# -*- coding: utf-8 -*-
from __future__ import (
    absolute_import,
    division,
    print_function,
    unicode_literals,
)

"""
Cookiecutter-Git Post Project Generation Hook Module.
"""
import base64
from contextlib import contextmanager
import errno
import getpass
import json
import os
import re
import shutil

if os.name == "nt":

    def quote(arg):
        # https://stackoverflow.com/a/29215357
        if re.search(r'(["\s])', arg):
            arg = '"' + arg.replace('"', r"\"") + '"'
        meta_chars = '()%!^"<>&|'
        meta_re = re.compile(
            "(" + "|".join(re.escape(char) for char in list(meta_chars)) + ")"
        )
        meta_map = {char: "^%s" % char for char in meta_chars}

        def escape_meta_chars(m):
            char = m.group(1)
            return meta_map[char]

        return meta_re.sub(escape_meta_chars, arg)


else:
    try:  # py34, py35, py36, py37
        from shlex import quote
    except ImportError:  # py27
        from pipes import quote

from invoke import Result, run, UnexpectedExit
import requests


class MockResult(Result):
    mock_stdout = {
        "github.com": """Password for 'https://NathanUrwin@github.com':
Counting objects: 18, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (15/15), done.
Writing objects: 100% (18/18), 6.24 KiB | 0 bytes/s, done.
Total 18 (delta 0), reused 0 (delta 0)
To https://github.com/NathanUrwin/cookiecutter-git-demo.git
* [new branch]      master -> master
Branch master set up to track remote branch master from origin.""",
        "gitlab.com": """Password for 'https://NathanUrwin@gitlab.com':
Counting objects: 13, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (10/10), done.
Writing objects: 100% (13/13), 5.25 KiB | 0 bytes/s, done.
Total 13 (delta 0), reused 0 (delta 0)
remote:
remote: The private project NathanUrwin/cookiecutter-git-demo was successfully created.
remote:
remote: To configure the remote, run:
remote:   git remote add origin https://gitlab.com/NathanUrwin/cookiecutter-git-demo.git
remote:
remote: To view the project, visit:
remote:   https://gitlab.com/NathanUrwin/cookiecutter-git-demo
remote:
To https://gitlab.com/NathanUrwin/cookiecutter-git-demo.git
* [new branch]      master -> master
Branch master set up to track remote branch master from origin.""",
        "bitbucket.org": """Password for 'https://NathanUrwin@bitbucket.org':
Counting objects: 13, done.
Delta compression using up to 8 threads.
Compressing objects: 100% (10/10), done.
Writing objects: 100% (13/13), 5.26 KiB | 0 bytes/s, done.
Total 13 (delta 0), reused 0 (delta 0)
To https://bitbucket.org/NathanUrwin/cookiecutter-git-demo.git
* [new branch]      master -> master
Branch master set up to track remote branch master from origin.""",
    }

    def __init__(self, remote_provider, **kwargs):
        stdout = self.mock_stdout.get(remote_provider, "")
        super(MockResult, self).__init__(stdout, **kwargs)


@contextmanager
def git_disable_gpgsign():
    """
    Disables git commit GPG signing temporarily.
    """
    try:
        result = run("git config --global --get commit.gpgSign", hide=True)
    # throws only on travis-ci. unknown reason
    except UnexpectedExit:

        class ResultNone:
            stdout = ""

        result = ResultNone()
    if result.stdout.strip() == "true":
        # turn off gpg signing commits
        try:
            run("git config --global --unset commit.gpgSign")
            yield
        # turn gpg signing back on
        except KeyboardInterrupt:
            run("git config --global --bool commit.gpgSign true")
        finally:
            run("git config --global --bool commit.gpgSign true")
    else:
        yield


class PostGenProjectHook(object):
    """
    Post Project Generation Class Hook.
    """

    bitbucket_repos_url_base = (
        "https://api.bitbucket.org/2.0/repositories/{}/{}"
    )
    create_remote_url = None
    github_repos_url = "https://api.github.com/user/repos"
    git_ignore_url_base = "https://www.gitignore.io/api/{}"
    headers = {}
    json_header = {"Content-Type": "application/json; charset=utf-8"}
    password_prompt_base = "Password for 'https://{}@{}': "
    remote_message_base = "Also see: https://{}/{}/{}"
    success_message_base = "\n\nSuccess! Your project was created here:\n{}\n{}\nThanks for using cookiecutter-git! :)\n\n"
    repo_dirpath = os.getcwd()
    cookiecutter_json_filepath = os.path.join(
        repo_dirpath, "cookiecutter.json"
    )
    raw_repo_slug_dirpath = os.path.join(
        repo_dirpath, "{% raw %}{{cookiecutter.repo_slug}}{% endraw %}"
    )
    github_dirpath = os.path.join(repo_dirpath, ".github")
    git_ignore_filepath = os.path.join(repo_dirpath, ".gitignore")
    hooks_dirpath = os.path.join(repo_dirpath, "hooks")
    licenses_dirpath = os.path.join(repo_dirpath, "LICENSES")
    license_filepath = os.path.join(repo_dirpath, "LICENSE")
    notice_filepath = os.path.join(repo_dirpath, "NOTICE")

    def __init__(self, *args, **kwargs):
        """
        Initializes the class instance.
        """
        self.result = self._get_cookiecutter_result()
        self.copyright_holder = self.result.get("copyright_holder")
        self.copyright_license = self.result.get("copyright_license")
        self.git_email = self.result.get("git_email")
        self.git_ignore = self.result.get("git_ignore")
        self.git_name = self.result.get("git_name")
        self.make_dirs = self.result.get("make_dirs")
        self.remote_namespace = self.result.get("remote_namespace")
        self.remote_protocol = self.result.get("remote_protocol")
        self.remote_provider = str(self.result.get("remote_provider")).lower()
        self.remote_username = self.result.get("remote_username")
        self.repo_slug = self.result.get("repo_slug")
        self.repo_summary = self.result.get("repo_summary")
        self.repo_tagline = self.result.get("repo_tagline")
        self._testing = str(self.result.get("_testing")).lower() == "true"
        #
        self.apache_license = self.copyright_license == "Apache-2.0"
        self.bitbucket_repos_url = self.bitbucket_repos_url_base.format(
            self.remote_namespace, self.repo_slug
        )
        self.git_ignore_url = self.git_ignore_url_base.format(self.git_ignore)
        self.nested_license_filepath = os.path.join(
            self.licenses_dirpath, "{}.txt".format(self.copyright_license)
        )
        self.new_git_ignore = self.git_ignore != "windows,macos,linux,git"
        self.password_prompt = self.password_prompt_base.format(
            self.remote_username, self.remote_provider
        )
        self.project_dirpaths = [
            os.path.join(self.repo_dirpath, dirname.strip())
            for dirname in self.make_dirs.split(",")
            if dirname.strip()
        ]
        self.remote_data = {
            "name": self.repo_slug,
            "description": self.repo_tagline,
        }
        self.create_remote_url = self._get_create_remote_url()
        self.encoded_data = json.dumps(self.remote_data).encode()
        self.remote_password = ""
        self.remote_repo = self.remote_provider != "none"
        if self.remote_repo:
            self.remote_password = self._get_remote_password()
        self.remote_origin_url = self._get_remote_origin_url()
        self.remote_origin_url_with_pass = self._get_remote_origin_url(
            with_pass=True
        )
        self.remote_message = (
            self.remote_message_base.format(
                self.remote_provider, self.remote_namespace, self.repo_slug
            )
            if self.remote_repo
            else ""
        )
        self.success_message = self.success_message_base.format(
            self.repo_dirpath, self.remote_message
        )

    @staticmethod
    def format_basic_auth(username, password):
        """
        Formats HTTP Basic auth with base64 encoding.

        :param username:
        :param password:
        """
        auth_data = "{}:{}".format(username, password)
        return base64.b64encode(auth_data.encode())

    @staticmethod
    def git_init():
        """
        Runs git init.
        """
        run("git init")

    @staticmethod
    def git_add():
        """
        Runs git add all.
        """
        # `git add -A`
        run("git add --all")

    @staticmethod
    def git_commit(message="Initial commit"):
        """
        Runs git commit with an initial message.

        :param message:
        """
        command = "git commit --message {}".format(quote(message))
        if os.name == "nt":
            # See https://github.com/NathanUrwin/cookiecutter-git/issues/43
            with git_disable_gpgsign():
                run(command)
        else:
            # `git commit -m "Initial commit"`
            run(command)

    @staticmethod
    def _get_cookiecutter_result():
        """
        Removes as much jinja2 templating from the hook as possible.
        """
        # http://flask.pocoo.org/docs/latest/templating/#standard-filters
        try:
            result = json.loads("""{{ cookiecutter | tojson() }}""")
        # current temp hack around for `pipenv run pytest -s`
        except json.JSONDecodeError:
            result = {}
            repo_dirpath = os.path.dirname(
                os.path.dirname(os.path.abspath(__file__))
            )
            json_filepath = os.path.join(repo_dirpath, "cookiecutter.json")
            with open(json_filepath) as f:
                for k, v in json.loads(f.read()).items():
                    result[k] = v
                    if isinstance(v, list):
                        result[k] = v[0]
        return result

    def _get_create_remote_url(self):
        """
        Gets the create remote URL.
        """
        create_remote_url = None
        if self.remote_username != self.remote_namespace:
            self.github_repos_url = "https://api.github.com/orgs/{}/repos".format(
                self.remote_namespace
            )
        if self.remote_provider == "bitbucket.org":
            create_remote_url = self.bitbucket_repos_url
            self.headers.update(self.json_header)
            self.remote_data.update({"has_issues": True, "is_private": True})
        elif self.remote_provider == "github.com":
            create_remote_url = self.github_repos_url
        return create_remote_url

    def _get_remote_origin_url(self, with_pass=False):
        """
        Gets the remote origin URL.
        """
        if with_pass:
            remote_origin_url = "https://{}:{}@{}/{}/{}.git".format(
                requests.utils.quote(self.remote_username, safe=""),
                requests.utils.quote(self.remote_password, safe=""),
                self.remote_provider,
                self.remote_namespace,
                self.repo_slug,
            )
        else:
            remote_origin_url = "https://{}@{}/{}/{}.git".format(
                requests.utils.quote(self.remote_username, safe=""),
                self.remote_provider,
                self.remote_namespace,
                self.repo_slug,
            )
        if self.remote_protocol == "ssh":
            remote_origin_url = "git@{}:{}/{}.git".format(
                self.remote_provider, self.remote_namespace, self.repo_slug
            )
        return remote_origin_url

    def _get_remote_password(self):
        """
        Gets the remote password with getpass.
        """
        # current hack for tests. better work around ?
        if self._testing:
            remote_password = "notmypassword"
        else:
            # prompt CLI users for password
            remote_password = getpass.getpass(self.password_prompt).strip()
        return remote_password

    def _git_remote_set(self):
        """
        Sets the git remote origin url.
        """
        run(
            "git remote set-url origin {}".format(
                quote(self.remote_origin_url)
            )
        )

    def _git_push(self):
        """
        Pushes the git remote and sets as upstream.
        """
        command = "git push --set-upstream origin master"
        if self._testing:
            print("_testing _git_push")
            r = MockResult(self.remote_provider, command=command)
            print(r.stdout)
            return r
        else:
            # `git push -u origin master`
            run(command)

    def _git_remote_add(self):
        """
        Adds the git remote origin url with included password.
        """
        run(
            "git remote add origin {}".format(
                quote(self.remote_origin_url_with_pass)
            )
        )

    def _git_remote_repo(self):
        """
        Creates a remote repo if needed.
        """
        if self.remote_provider in ["bitbucket.org", "github.com"]:
            # http basic auth if needed
            auth_base64 = self.format_basic_auth(
                self.remote_username, self.remote_password
            )
            self.headers["Authorization"] = "Basic {}".format(
                auth_base64.decode()
            )
            # create remote POST request
            requests.post(
                self.create_remote_url,
                data=self.encoded_data,
                headers=self.headers,
            )

    def _git_ignore(self):
        """
        Creates a new .gitignore if needed from gitignore.io.
        """
        if self.new_git_ignore:
            # gitignore.io/api -> {{cookiecutter.repo_slug}}/.gitignore
            response = requests.get(self.git_ignore_url)
            with open(self.git_ignore_filepath, "w") as f:
                f.write(response.text)

    def _git_repo(self):
        """
        Adds a .gitignore, initial commit, and remote repo.
        """
        self._git_ignore()
        self.git_init()
        self.git_add()
        self.git_commit()
        if self.remote_repo:
            self._git_remote_repo()
            self._git_remote_add()
            self._git_push()
            self._git_remote_set()

    def _github_dir(self):
        """
        Removes the .github dir if GitHub is not the remote provider.
        """
        if not self.remote_provider == "github.com":
            shutil.rmtree(self.github_dirpath, ignore_errors=True)

    def _make_dirs(self):
        """
        Makes dirs and adds .gitkeep files to them.
        """
        for dirpath in self.project_dirpaths:
            # equal to bash `mkdir -p $dirpath`
            try:
                os.makedirs(dirpath)
            except OSError as exc:
                if exc.errno == errno.EEXIST and os.path.isdir(dirpath):
                    pass
                else:
                    raise
            # equal to bash `touch $dirpath/.gitkeep`
            gitkeep = os.path.join(dirpath, ".gitkeep")
            with open(gitkeep, "a"):
                os.utime(gitkeep, None)

    def _copyright_license(self):
        """
        Adds the chosen LICENSE and removes the rest.
        """
        if not self.apache_license:
            os.remove(self.notice_filepath)
        shutil.move(self.nested_license_filepath, self.license_filepath)
        shutil.rmtree(self.licenses_dirpath, ignore_errors=True)

    def run(self):
        """
        Sets up the project license, dirs, and git repo.
        """
        self._copyright_license()
        self._make_dirs()
        self._github_dir()
        self._git_repo()
        print(self.success_message)


def main():
    """
    Runs the post gen project hook main entry point.
    """
    PostGenProjectHook().run()


# This is required! Don't remove!!
if __name__ == "__main__":
    main()
